Modules/permissions.py

```diff
@@ -24,23 +26,140 @@ class Permission:
     A permission level
     """
 
-    def __init__(self, level: int, vhost: str,
+    def __init__(self, level: int, vhosts: Set[str],
                  deny_message: str = "Access denied."):
         """
-        Permission required to execute a command
-        :param level: Relative permissions level
-        :param vhost: associated vhost
-        :param deny_message: message to display if user level < level required
-        :return:
+        creates a representation of a permission level required to execute an IRC command
+
+        Args:
+            level (Permission): required Permission level
+            vhosts (Set[str]): set of vhost strings that fall under this permission level
+            deny_message (str): message displayed on message to display if user level < level
+                required
         """
+
         log.debug(f"Created new Permission object with permission {level}")
         self._level = level
-        self._vhost = vhost
+        self._vhosts = vhosts
         self._denied_message = deny_message
 
-    level = property(lambda self: self._level)
-    vhost = property(lambda self: self._vhost)
-    denied_message = property(lambda self: self._denied_message)
+        _by_vhost.update({vhost: self for vhost in self._vhosts})
+
+    @property
+    def level(self) -> int:
+        """
+        Permission level
+
+        Returns:
+            int
+        """
+        return self._level
+
+    @property
+    def vhosts(self) -> Set[str]:
+        """
+        Registered vhosts that fall under this permission level
+
+        Returns:
+            List[str]: list of registered vhosts
+        """
+        return self._vhosts
+
+    @vhosts.setter
+    def vhosts(self, value: Set[str]) -> None:
+        """
+        Updates the registered vhosts for this permission object
+
+        This also has the side effect of updating the _by_vhost bindings for this object.
+
+        - Items removed from `vhosts` are removed from _by_vhost and
+        - Items that are not in `vhosts` already are added to `_by_vhost`
+
+        Args:
+            value (List[str]): list of vhosts
+
+        Returns:
+            None
+        """
+        if not isinstance(value, set):
+            raise TypeError(f"expected list got {type(value)}")
+
+        # determine which items have been removed
+        removed = self.vhosts - value
+
+        # determine which items are new
+        added = value - self.vhosts
+
+        # recompute _by_vhost to exclude removed vhosts we have removed
+        # ( as there is no bulk-remove method, this avoids a for loop and some function calls )
+        all_vhosts = {vhost: permission for vhost, permission in _by_vhost.items() if
+                      vhost not in removed}
+
+        # create a dict of new vhosts to register
+        new_vhosts = {vhost: self for vhost in added}
```

Comment:
It will probably be safer, simpler and faster to just build up `_by_vhost` from scratch.

Reply:
`_by_vhost` is being rebuilt from scratch. 
The line you cite there is where im building a dict of all new permissions being registered, which gets added to `all_vhosts` which is `_by_vhost` stripped of all vhosts that are being deregistered.

`all_vhosts` gets combined with `new_vhosts` which is then pushed to `_by_vhost`

Reply:
I'm only noticing your response now, sorry. Looks like I didn't get an email.
You're right, I misread part of the code, it is necessary to rely on the old `_by_vhost`, although it could be simplified with something along the lines of:
```python
global _by_vhost
_by_vhost = {vhost: permission for vhost, permission in _by_vhost.items() if permission is not self or vhost}  # all vhosts that aren't ours
_by_vhost.update((vhost, self) for vhost in value)  # ours
```
But that's just optional isn't it.

Reply:
_by_vhost is a dict, and you appear to be suggesting a tupple comprehension? 

at any rate i think it is desirable to keep the dict comprehension as its own step.

Reply:
`dict.update` has an overload `(Iterable[tuple])`, but as I said your version is fine